src/Microsoft.DotNet.Interactive/RoutingSlip.cs

```diff
@@ -11,49 +12,103 @@ namespace Microsoft.DotNet.Interactive;
+            _uris = new List<Entry>();
         }
     }
 
-    public bool TryAdd(Uri kernelOrKernelHostUri)
+    public bool TryMarkArrival(Uri kernelOrKernelHostUri)
     {
-        lock (_lock)
+        return TryMark(kernelOrKernelHostUri, false);
+    }
+
+    public bool TryMarkHandled(Uri kernelOrKernelHostUri)
+    {
+        return TryMark(kernelOrKernelHostUri, true);
+    }
+
+    private bool TryMark(Uri kernelOrKernelHostUri, bool handled)
+    {
+        var entry = new Entry(kernelOrKernelHostUri)
         {
-            if (_uniqueUris.Add(kernelOrKernelHostUri))
+            Handled = handled
+        };
+        var ret = _entriesByUri.TryAdd(kernelOrKernelHostUri, entry);
+        if (ret)
+        {
+            _uniqueUris.Add(kernelOrKernelHostUri);
+            _uris.Add(new Entry(kernelOrKernelHostUri));
+        }
+
+        return ret;
+    }
+
+    public void MarkHandled(Uri kernelOrKernelHostUri)
+    {
+        if (_entriesByUri.TryGetValue(kernelOrKernelHostUri, out var entry))
+        {
+            if (entry.Handled)
+            {
+                throw new InvalidOperationException($"The routing slip has already been handled by {kernelOrKernelHostUri}");
+            }
+            else
             {
-                _uris.Add(kernelOrKernelHostUri);
-                return true;
+                entry.Handled = true;
             }
         }
-
-        return false;
+        else
+        {
+            throw new InvalidOperationException(
+                $"The kernel or kernel host {kernelOrKernelHostUri} is not part of this routing slip");
+        }
     }
 
-    public bool Contains(Uri kernelOrKernelHostUri) => _uniqueUris.Contains(kernelOrKernelHostUri);
+    public bool Contains(Uri kernelOrKernelHostUri) => _entriesByUri.ContainsKey(kernelOrKernelHostUri);
 
     public bool Contains(RoutingSlip other)
     {
-        var contains =  this.Zip(other, (o, i) => o.Equals(i)).All(x => x);
+        var contains_old = this.Zip(other, (o, i) => o.Equals(i)).All(x => x);
+        var contains = _uniqueUris.IsSupersetOf(other._uniqueUris);
         return contains;
     }
 
-    IEnumerator<Uri> IEnumerable<Uri>.GetEnumerator() => _uris.GetEnumerator();
+    IEnumerator<Uri> IEnumerable<Uri>.GetEnumerator() => _uris.Select(e => e.KernelUri).GetEnumerator();
 
     IEnumerator IEnumerable.GetEnumerator() => _uris.GetEnumerator();
 
-    public Uri this[int index] => _uris[index];
+    public Uri this[int index] => _uris[index].KernelUri;
     public int Count => _uris.Count;
+
+    public class Entry
+    {
+        public Entry(Uri kernelUri)
+        {
+            KernelUri = kernelUri;
+            Handled = false;
+        }
+
+        public bool Handled { get; internal set; }
+
+        public Uri KernelUri { get; }
+    }
+
+    public bool HasProcessed(Uri kernelInfoUri)
+    {
+        return _entriesByUri.TryGetValue(kernelInfoUri, out var entry) && entry.Handled;
```

Comment:
It's better to use consistent vocabulary, and there's always a better word than "processed".